core/tests/ledger_cleanup.rs

```diff
@@ -338,27 +338,52 @@ mod tests {
                     let mut num_shreds = 0;
                     let mut max_speed = 0f32;
                     let mut min_speed = f32::MAX;
+                    let (mut shreds_with_parent, _) = make_many_slot_shreds(
+                        1, batch_size_slots, shreds_per_slot);
+                    let (first_shreds, _) = make_many_slot_shreds(
+                        0, batch_size_slots, shreds_per_slot);
                     loop {
-                        let batch_id = shared_batch_id.fetch_add(1, Ordering::SeqCst);
+                        let batch_id = shared_batch_id.fetch_add(1, Ordering::Relaxed);
                         let start_slot = batch_id * batch_size_slots;
-                        let len = batch_id;
                         if start_slot >= benchmark_slots {
                             break;
                         }
+                        let len = batch_id;
 
-                        let new_shreds = if pre_generate_data {
+                        let br = if pre_generate_data {
                             let mut sl = cloned_shreds.lock().unwrap();
                             if let Some(shreds_from_queue) = sl.pop_front() {
-                                shreds_from_queue
+                                total += shreds_from_queue.len();
+                                cloned_blockstore.insert_shreds(
+                                    shreds_from_queue, None, false).unwrap()
                             } else {
+                                // If the queue is empty, we're done!
                                 break;
                             }
                         } else {
-                            let (generated_shreds, _) = make_many_slot_shreds(
-                                start_slot, batch_size_slots, shreds_per_slot);
-                            generated_shreds
+                            let mut slot_id = start_slot;
+                            if slot_id > 0 {
+                                for shred in shreds_with_parent.iter_mut() {
+                                    shred.set_slot(slot_id);
```

Comment:
@yhchiang-sol Can you change the code here not to use `.set_slot`?
I have to remove that api because it invalidates shreds, and breaks other things. See https://github.com/solana-labs/solana/pull/27689

This test is failing in https://github.com/solana-labs/solana/pull/27675 because of `.set_slot` here.

Reply:
Sure thing.  Let me see if there's any workaround.

Reply:
@behzadnouri: this part in the benchmark code is for quickly generating shreds by cloning and modifying important fields (in this case, the slot_id) instead of re-generating the whole shreds (otherwise, the shred generation actually runs slower than shred insertion, which breaks the purpose of the shred insertion benchmark).  For this, I think I might need to move the function to the test file itself.

For ShredVariant::MerkleCode and ShredVariant::MerkleData, can I know how this should be implemented?

Reply:
I made some changes in https://github.com/solana-labs/solana/pull/27675 to make the test pass.
Please take a look if the change makes sense.